Make Terraformer access to IAM read-only

jsf9k opened the pull request:
## 🗣 Description ##

This pull requests adds a statement to the Terraformer policy document that makes its IAM access read-only.

## 💭 Motivation and context ##

This stops the Terraformer instance from defeating the Terraformer policy by creating new users, policies, roles, etc.  I'm looking at you, @epicfaace!  :smiley:

## 🧪 Testing ##

I successfully applied these changes to env1 in our staging COOL environment.

## ✅ Checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

jsf9k committed:
Make Terraformer access to IAM read-only

This stops the Terraformer instance from defeating the Terraformer
policy by creating new users, policies, roles, etc.

dav3r reviewed (approved):
Makes sense to me! 🔒

mcdonnnj commented on terraformer_policy.tf:
```diff
@@ -28,6 +28,23 @@ data "aws_iam_policy_document" "terraformer_policy_doc" {
+  # creating new users, policies, roles, etc.
+  #
+  # I cribbed this list of actions from the AWS built-in policy
+  # arn:aws:iam::aws:policy/IAMReadOnlyAccess.
```
I think we should mention the rationale for not including the `iam:Generate*` permission from the built-in policy. An alternative would be to link to [this page in the AWS documentation](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_examples_iam_read-only-console-no-reporting.html) since it matches the permissions we're allowing.

jsf9k replied:
I believe this outdated conversation can now be resolved.

mcdonnnj reviewed (changes requested):
Makes sense. I had one request related to clarifying the documentation around the new policy statement.

jsf9k committed:
Refine IAM permissions

In testing I realized that my previous method of clamping down on IAM
permissions was too harsh.  We have to allow the iam;PassRole
permission for spinning up instances and assigning them instance
roles.

dav3r commented on terraformer_policy.tf:
```diff
@@ -11,22 +11,58 @@ data "aws_iam_policy_document" "terraformer_policy_doc" {
+  # root module, with the exception of IAM.
+  #
+  # We will attach the arn:aws:iam::aws:policy/ReadOnlyAccess policy,
+  # which will cover the IAM case.  We require IAM access to be
```
This comment now needs to be updated.

jsf9k replied:
That policy gets attached [here](https://github.com/cisagov/cool-assessment-terraform/blob/93c422af1983953f184a70a7f2f5f5000747da87/terraformer_role.tf#L13-L19).  What needs to change?

jsf9k replied:
Have a look at commit 041d892 and let me know if that scratches your itch.

dav3r replied:
> That policy gets attached [here](https://github.com/cisagov/cool-assessment-terraform/blob/93c422af1983953f184a70a7f2f5f5000747da87/terraformer_role.tf#L13-L19). What needs to change?

Arg, I totally overlooked that- this is why I shouldn't review PRs at the end of the day.  Though the enhanced comments in https://github.com/cisagov/cool-assessment-terraform/commit/041d892a59aaa6bafa6a6b1fa0ccede95837f0b8 are lovely. 👍

jsf9k replied:
No problem sir!

dav3r reviewed (approved):
Other than the comment change that I noted, this all looks good to me. 🦘

jsf9k committed:
Enhance comments in Terraformer policy document

dav3r commented on terraformer_policy.tf:
```diff
@@ -11,22 +11,60 @@ data "aws_iam_policy_document" "terraformer_policy_doc" {
+  #
+  # We will attach the arn:aws:iam::aws:policy/ReadOnlyAccess policy
+  # to the same role to which policy document will be attached, which
+  # will give is read-only access to all IAM resources.  We require
```
Fix a couple of typos:
```suggestion
  # We will attach the arn:aws:iam::aws:policy/ReadOnlyAccess policy
  # to the same role to which the policy document will be attached, which
  # will give it read-only access to all IAM resources.  We require
```

jsf9k committed:
Fix a couple of typos

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj reviewed (approved):
This makes sense. 🔒
